Add ADR for one database per state decision (#2462)
18F/piipan: docs/adr/0022-have-one-database-per-state.md

We believe the benefits of this approach largely dissolved once we pivoted to our [PPRL approach](../pprl.md), where deidentified PII from states is stored in the NAC and matches are performed directly against this deidentified data.

We understand that having a separate NAC Postgresql database for each state now brings added complexity to the project without the original intended benefits. This opens up new possibilities for how to store this data more effectively. In deciding what will be best going forward, we believe the following considerations are still important:

1. Using the principle of least privilege: states should have no access to other states' databases under any circumstances
1. Authorization: states should not be able to modify their data once they've submitted it to the NAC. (In the current system, they can replace their data, but not modify or delete it.) There also may be no reason for states to read their own data once it's submitted.
1. Security: how would the new system change the threat model to the data?

> 1. Scalability: the future data store may contain hundreds of millions of records at a time

Reviewer: ```suggestion
1. Scalability: the future data store will contain at least tens of millions of records at a time
```

> 1. Performance: how can data be stored to make realtime matching faster?

Reviewer: ```suggestion
1. Performance: how can data be stored to make near real-time matching faster?
```

> 1. That it's a well-supported solution within the Microsoft/Azure ecosystem

Reviewer: ```suggestion
1. Well established support within the Microsoft/Azure ecosystem
```

1. Ease of use for the [Bulk Upload API](../openapi/generated/bulk-api/openapi.md)

## Decision

We will keep our current database structure while remaining open to possibilities for change. Making these changes will be weighed against other competing project priorities.

## Consequences

> * We will need to support and maintain a separate Postgresql database for each state and ensure schema consistency across all of them.

Reviewer: ```suggestion
* We will need to support and maintain a separate PostgreSQL database for each state and ensure schema consistency across all of them.
```
